src/openfermion/ops/_symbolic_operator.py

```diff
@@ -96,13 +96,64 @@ def __init__(self, term=None, coefficient=1.):
         # Add the term to the dictionary
         self.terms[term] = coefficient
 
-    def _long_string_init(self, term, coefficient):
+    def _long_string_init(self, long_string, coefficient):
         """
-        Initialization from a long string representation, i.e., term
-        is a string such as '1.5 [2^ 3] + 2.4 [3^ 0]'.
+        Initialization from a long string representation.
+
+        e.g. For FermionOperator:
+            '1.5 [2^ 3] + 1.4 [3^ 0]'
         """
-        raise NotImplementedError('Initializing a SymbolicOperator from a '
-                                  'long string is not yet supported.')
+
+        index = 0  # Index of current character in long_string
+        subtraction = False  # Whether this term is being subtracted
+        while index < len(long_string):
+
+            # Find the start and end indices of the next coefficient
+            while long_string[index].isspace():
+                index += 1
+            coefficient_start = index
+            while long_string[index] is not '[':
+                index += 1
+            coefficient_end = index
+
+            # Find the start and end indices of the term
+            factor_start = index + 1
+            while long_string[index] is not ']':
+                index += 1
+            factor_end = index
+
+            # Determine the coefficient for this term
+            if coefficient_start == coefficient_end:
+                coef = 1.0
+            else:
+                try:
+                    coef = float(
+                            long_string[coefficient_start:coefficient_end])
+                except ValueError:
+                    raise ValueError('Invalid coefficient {}.'.format(
+                            long_string[coefficient_start:coefficient_end]))
+            coef *= coefficient
+            if subtraction:
+                coef = -coef
```

Comment:
I think it is slightly nicer to write "coef *= -1". I don't feel too strongly though.